plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js

```diff
@@ -3,79 +3,376 @@ import {
+        break;
+
+      case 'gracePeriodSeconds':
+        newState.gracePeriodSeconds = intOrNull(value);
+        break;
+
+      case 'intervalSeconds':
+        newState.intervalSeconds = intOrNull(value);
+        break;
+
+      case 'maxConsecutiveFailures':
+        newState.maxConsecutiveFailures = intOrNull(value);
+        break;
+
+      case 'timeoutSeconds':
+        newState.timeoutSeconds = intOrNull(value);
+        break;
+
+      case 'delaySeconds':
+        newState.delaySeconds = intOrNull(value);
+        break;
+    }
+
+    return newState;
+  },
+
+  JSONSegmentParser(healthCheck, path) {
+    const memo = [];
+
+    // Parse detailed fields according to type
+    if (healthCheck.http != null) {
+      memo.push(new Transaction(path.concat(['protocol']), HTTP, SET));
+      parseHttpHealthCheck(healthCheck.http, path.concat(['http']), memo);
+    }
+    if (healthCheck.tcp != null) {
+      memo.push(new Transaction(path.concat(['protocol']), TCP, SET));
+      parseTcpHealthCheck(healthCheck.tcp, path.concat(['tcp']), memo);
+    }
+    if (healthCheck.exec != null) {
+      memo.push(new Transaction(path.concat(['protocol']), COMMAND, SET));
+      parseCommandHealthCheck(healthCheck.exec, path.concat(['exec']), memo);
+    }
+
+    // Parse generic fields
+    if (healthCheck.gracePeriodSeconds != null) {
+      memo.push(new Transaction(
+        path.concat(['gracePeriodSeconds']),
+        parseInt(healthCheck.gracePeriodSeconds),
+        SET
+      ));
+    }
+    if (healthCheck.intervalSeconds != null) {
+      memo.push(new Transaction(
+        path.concat(['intervalSeconds']),
+        parseInt(healthCheck.intervalSeconds),
+        SET
+      ));
+    }
+    if (healthCheck.maxConsecutiveFailures != null) {
+      memo.push(new Transaction(
+        path.concat(['maxConsecutiveFailures']),
+        parseInt(healthCheck.maxConsecutiveFailures),
+        SET
+      ));
+    }
+    if (healthCheck.timeoutSeconds != null) {
+      memo.push(new Transaction(
+        path.concat(['timeoutSeconds']),
+        parseInt(healthCheck.timeoutSeconds),
+        SET
+      ));
+    }
+    if (healthCheck.delaySeconds != null) {
+      memo.push(new Transaction(
+        path.concat(['delaySeconds']),
+        parseInt(healthCheck.delaySeconds),
+        SET
+      ));
+    }
+
+    return memo;
+  },
+
+  FormReducer(state, {type, path, value}) {
+    const newState = MultiContainerHealthChecks.JSONSegmentReducer
+      .call(this, state, {type, path, value});
+
+    // Bail early on nulled cases
+    if (newState == null) {
+      return newState;
+    }
+
+    const [group, field, secondField] = path;
+
+    // Include additional fields only present in the form
+    if (group === 'protocol') {
+      newState.protocol = value;
+    }
+
+    // Assign detailed properties
+    switch (group) {
+      case 'exec':
+        reduceFormCommandHealthCheck(newState, secondField, value);
```

Comment:
I'd rather not mutate newState here
then we could even `return reduceFormCommandHealthCheck(...)` right away